File: leetcode 907.py
from typing import List
class Solution:
    M = 1e9 + 7
    def sumSubarrayMins(self, arr: List[int]) -> int:
        arr.append(-1)
        n = len(arr)
        stk = [-1]
        ans = 0
        for i in range(0, n):
            while len(stk) > 1 and arr[stk[-1]] >= arr[i]:
                center = stk.pop()
                left = stk[-1]; right = i
                ans = (ans + arr[center] * (right - center) * (center - left)) % self.M
            stk.append(i)
        return int(ans)

File: leetcode 907.cpp
class Solution {
public:
  int sumSubarrayMins(vector<int> &arr) {
    arr.push_back(-1); // padding
    int n = arr.size(), ans = 0;
    vector<int> stk;
    stk.push_back(-1); // padding
    for (int i = 0; i < n; i++) {
      while (stk.size() > 1 && arr[stk.back()] >= arr[i]) {
        int center = stk.back();
        stk.pop_back();
        int left = stk.back(), right = i;
        ans = (static_cast<long long>(ans) +
               arr[center] * (center - left) % M *
                   static_cast<long long>(right - center)) %
              M;
        // 因为M是1e9级别，所以不加static_cast的话，乘法会溢出int
      }
      stk.push_back(i);
    }
    return ans;
  }
};